source/run/src/TG4PrimaryGeneratorAction.cxx

```diff
@@ -36,10 +40,26 @@
 // generated from short units names
 #include <G4SystemOfUnits.hh>
 
+namespace
+{
+
+TG4EventAction* GetEventAction()
+{
+  return dynamic_cast<TG4EventAction*>(const_cast<G4UserEventAction*>(
```

Comment:
This code is problematic for two reasons:
1) dynamic cast is slow
2) dynamic cast can return 0 and if this happens, the program will break
Please, take a look at  the pattern in eg. TG4SteppingAction::ProcessTrackOnBoundary, where dynamic_cast is used only under MCDEBUG option including a test of result, and there is used a static cast otherwise

Reply:
Agreed.
However, in a case where a `dynamic_cast` would return a `nullptr`, `static_cast` would of course also fail since what it returns is not valid. But on the other hand, the point where the the `TG4EventAction` is extracted, there would be a deeper problem, if it wasn't present.
Finally, with a `static_cast` should be fine.